pkg/nodeagent/controller/operatingsystemconfig/containerd.go

```diff
@@ -0,0 +1,363 @@
+	)
+
+	for _, registryConfig := range newRegistries {
+		if ptr.Deref(registryConfig.ReadinessProbe, false) {
+			registriesWithReadiness = append(registriesWithReadiness, registryConfig)
+		}
+		registriesWithoutReadiness = append(registriesWithoutReadiness, registryConfig)
+	}
+
+	// Registries without readiness probes can directly and synchronously be added here
+	// since there is no longer blocking operation involved.
+	for _, registryConfig := range registriesWithoutReadiness {
+		if err := addRegistryToContainerdFunc(ctx, log, registryConfig, r.FS); err != nil {
+			errChan <- err
+			return errChan
+		}
+	}
+
+	fns := make([]flow.TaskFn, 0, len(registriesWithReadiness))
+	for _, registryConfig := range registriesWithReadiness {
+		fns = append(fns, func(ctx context.Context) error {
+			return addRegistryToContainerdFunc(ctx, log, registryConfig, r.FS)
+		})
+	}
+
+	go func() {
+		errChan <- flow.Parallel(fns...)(ctx)
+	}()
+
+	return errChan
+}
+
+func addRegistryToContainerdFunc(ctx context.Context, log logr.Logger, registryConfig extensionsv1alpha1.RegistryConfig, fs afero.Afero) error {
+	httpClient := http.Client{Timeout: 1 * time.Second}
+
+	baseDir := path.Join(certsDir, registryConfig.Upstream)
+	if err := fs.MkdirAll(baseDir, defaultDirPermissions); err != nil {
+		return fmt.Errorf("unable to ensure registry config base directory: %w", err)
+	}
+
+	hostsTomlFilePath := path.Join(baseDir, "hosts.toml")
+	exists, err := fs.Exists(hostsTomlFilePath)
+	if err != nil {
+		return fmt.Errorf("unable to check if registry config file exists: %w", err)
+	}
+
+	// Check if registry endpoints are reachable if the config is new.
+	// This is especially required when registries run within the cluster and during bootstrap,
+	// the Kubernetes deployments are not ready yet.
+	if !exists && ptr.Deref(registryConfig.ReadinessProbe, false) {
+		log.Info("Probing endpoints for image registry", "upstream", registryConfig.Upstream)
+		if err := retry.Until(ctx, 2*time.Second, func(ctx context.Context) (done bool, err error) {
+			for _, registryHosts := range registryConfig.Hosts {
+				req, err := http.NewRequestWithContext(ctx, http.MethodGet, registryHosts.URL, nil)
+				if err != nil {
+					return false, fmt.Errorf("failed to construct http request %s for upstream %s: %w", registryHosts.URL, registryConfig.Upstream, err)
+				}
+
+				_, err = httpClient.Do(req)
+				if err != nil {
+					return false, fmt.Errorf("failed to reach registry %s for upstream %s: %w", registryHosts.URL, registryConfig.Upstream, err)
+				}
+			}
+			return true, nil
+		}); err != nil {
+			return err
+		}
+
+		log.Info("Probing endpoints for image registry succeeded", "upstream", registryConfig.Upstream)
+	}
+
+	f, err := fs.OpenFile(hostsTomlFilePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("unable to open hosts.toml: %w", err)
+	}
+
+	defer func() {
+		err = f.Close()
+	}()
```

Comment:
If err != nil, err won't be returned to the caller. Not sure if this is expected.